homeassistant/components/zwave_js/services.py

```diff
@@ -0,0 +1,216 @@
+"""Methods and classes related to executing Z-Wave commands and publishing these to hass."""
+import logging
+from typing import Dict, List, Union, cast
+
+import voluptuous as vol
+from zwave_js_server.const import CommandClass
+from zwave_js_server.model.node import Node as ZwaveNode, get_value_id
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_DEVICE_ID, ATTR_ENTITY_ID
+from homeassistant.core import HomeAssistant, ServiceCall, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.device_registry import DeviceRegistry
+from homeassistant.helpers.entity_registry import EntityRegistry
+
+from . import const
+from .entity import get_home_and_node_id_from_device_id
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def convert_bitmask_to_int(value: str) -> int:
+    """Convert a bitmask (hex) to an integer."""
+    return int(value, 16)
+
+
+def parameter_name_does_not_need_bitmask(
+    val: Dict[str, Union[int, str]]
+) -> Dict[str, Union[int, str]]:
+    """Validate that if a parameter name is provided, bitmask is not as well."""
+    if isinstance(val[const.ATTR_CONFIG_PARAMETER], str) and (
+        val.get(const.ATTR_CONFIG_PARAMETER_BITMASK)
+    ):
+        raise vol.Invalid(
+            "Don't include a bitmask when a parameter name is specified",
+            path=[const.ATTR_CONFIG_PARAMETER, const.ATTR_CONFIG_PARAMETER_BITMASK],
+        )
+    return val
+
+
+BITMASK_SCHEMA = vol.All(
+    cv.string, vol.Lower, vol.Match(r"^(0x)?[0-9a-f]+$"), convert_bitmask_to_int
+)
+
+
+class ZWaveServices:
+    """Class that holds our services (Zwave Commands) that should be published to hass."""
+
+    def __init__(
+        self, hass: HomeAssistant, dev_reg: DeviceRegistry, ent_reg: EntityRegistry
+    ):
+        """Initialize with hass, device registry, and entity registry objects."""
+        self._hass = hass
+        self._dev_reg = dev_reg
+        self._ent_reg = ent_reg
+
+    @callback
+    def async_register(self) -> None:
+        """Register all our services."""
+        self._hass.services.async_register(
+            const.DOMAIN,
+            const.SERVICE_SET_CONFIG_PARAMETER,
+            self.async_set_config_value,
+            schema=vol.All(
+                {
+                    vol.Exclusive(ATTR_DEVICE_ID, "id"): cv.string,
+                    vol.Exclusive(ATTR_ENTITY_ID, "id"): cv.entity_ids,
+                    vol.Optional(const.ATTR_ENDPOINT): vol.Coerce(int),
+                    vol.Required(const.ATTR_CONFIG_PARAMETER): vol.Any(
+                        vol.Coerce(int), cv.string
+                    ),
+                    vol.Optional(const.ATTR_CONFIG_PARAMETER_BITMASK): vol.Any(
```

Comment:
Can't we make this smarter by just always accepting value ?
We can determine the type from the value and we know what the device supports....

Reply:
are you talking about the `bitmask` or the `value` parameter of the input data? there are three ways to enter a `bitmask`: it's integer representation, its bitmask with 0x preceding it, and its bitmask without 0x preceding it. The validation ensures that it is in one of these formats and converts it to its integer representation because thats the format the data is in.

Reply:
I mean that we could move that logic into the lib and let the user just give a value in any of the supported formats.
The ATTR_CONFIG_PARAMETER_BITMASK is implying that you need a different argument for bitmask value types.
Or am I reading it wrong ?

Reply:
What I mean is that the "value" parameter for the service call can be of any of the supported types (string, int or dict) and we handle parsing.

```
service.set_config_param:
  entity_id: blah
  parameter: 1
  value: 1
```

or

```
service.set_config_param:
  entity_id: blah
  parameter: Some label that describes the config param
  value: blah
```

or

```
service.set_config_param:
  entity_id: blah
  parameter: Some label that describes the config param
  value:
    1: blah
    2: blah
```


or

```
service.set_config_param:
  entity_id: blah
  parameter: Some label that describes the config param
  value:
    'Cool configuration value': blah
```

Reply:
So, "parameter" can be of type string (the label) or int
"value" can be of type string, label, Dict[str, int] or Dict[int,int] or Dict[int, str] or Dict[str,str]

But maybe restrict it a tiny bit and either allow the key as int and value as int or both strings for the dict types

Reply:
right so the one you are missing is parameter + bitkey (Scenario 1 in the one I listed above). For partial bits you would need to supply two numbers to get the right bit

Reply:
OK, but in that case the user has to know that the value is a bitmask value, right ?
From what I understand of it the parameter will be the propertyKey and the dict on the value will be the various bitmask values.

So we abstract out the whole bitmask stuff for the user as most likely the config UI will also not display that to the user and just provide labels for all values.
But mabe I'm overseeing something or trying to go a bridge too far here :-)

Reply:
for use cases where a user wants to update a partial value, they have two options with this implementation:
1. Provide property (parameter) and propertyKey (bitmask with an automatic conversion from the format in the device DB to the format that the library understands)
2. Provide the property name as the parameter

For use cases where a user wants to update multiple partial values, they will also have two options:
1. Provide a key/value mapping in the value where the key is either the properteyKey or the property name
2. Provide a raw int value which gets converted into the appropriate partial values

Neither of these cases for updating multiple partial values can be supported right now, so they are not a part of the schema

Reply:
Here's what the schema will eventually look like for value when we do support updating multiple partial values:
```python
vol.Required(const.ATTR_CONFIG_VALUE): vol.Any(
    vol.Coerce(int),
    cv.string,
    {
        vol.Any(
            vol.Coerce(int), BITMASK_SCHEMA, cv.string
        ): vol.Any(vol.Coerce(int), cv.string)
    },
),
```

Reply:
and the plan there is to pass the data as is to the library where it can do the work of converting things into a sendable value

Reply:
all clear now! So the gist of this is:
the MVP will only support setting values one by one and a later iteration might support setting multiple (bitmask) values all at once.

I assume there is also the usecase of simple values without the bitmask stuff ?

And the config UI will break out each individual bitmask value into its own configuration setting, right ?
The label would be something like "propertyName + propertyKeyName"

Reply:
Yes, this supports the usecase of config parameters that dont have bitmasks as well.

Re: UI that is a better question for @cgarwood but I think that would make sense. We just need to ensure that the UI allows for multiple bits to be set at the same time to handle the scenarios petro described in Discord yesterday

Reply:
Yeah, UI will break out the bitmasks into separate properties grouped together